### PR TITLE
Fairscale FSDP fix model save

### PR DESCRIPTION
**Pull request by sgugger**
# What does this PR do?

This PR fixes the fact training with fairscale fully-sharded wrapper was hanging: it looks like recent changes in fairscale make a synchronization during the model state dict call, which results in the training hanging if we don't call that state_dict method on all processes. This PR addresses that.

### CODE REVIEW

File: examples/tests/trainer/test_trainer_ext.py [hunk 1]
```diff
@@ -66,7 +66,7 @@ def require_apex(test_case):
 
 
 class TestTrainerExt(TestCasePlus):
-    def run_seq2seq_quick(self, distributed=False, extra_args_str=None, eval=True, predict_with_generate=True):
```
Comment (sgugger) [inline examples/tests/trainer/test_trainer_ext.py:69]: eval was not used anywhere.

File: src/transformers/trainer.py [hunk 2]
```diff
@@ -1495,11 +1495,14 @@ def save_model(self, output_dir: Optional[str] = None):
...
-        else:
+        elif (
+            ShardedDDPOption.ZERO_DP_2 in self.args.sharded_ddp or ShardedDDPOption.ZERO_DP_3 in self.args.sharded_ddp
+        ):
+            state_dict = self.model.state_dict()
             if self.is_world_process_zero():
-                self._save(output_dir)
-            if self.args.local_rank != -1:
-                dist.barrier()
+                self._save(output_dir, state_dict=state_dict)
```
Comment (sgugger) [inline src/transformers/trainer.py:1503]: Gather state dict on all processes (because of the sync) but save it on process zero only.

**Review by stas00 — APPROVED** [review]
It works now! 

Thank you for the fix and cleaning up my mistakes!

**Review by LysandreJik — APPROVED** [review]
LGTM!

**Pull request merged**